File: src/pkg01_topic_cpp/src/subscriber_node.cpp
// subscriber_node.cpp
// A minimal ROS2 subscriber in C++.
// Subscribes to the topic "/chatter" and prints every message received.

#include <memory>

#include "rclcpp/rclcpp.hpp"
#include "std_msgs/msg/string.hpp"

class SubscriberNode : public rclcpp::Node
{
public:
  SubscriberNode()
  : Node("subscriber_node")
  {
    // Create a subscriber:
    //   topic name : "chatter"  (must match the publisher!)
    //   QoS depth  : 10
    //   callback   : topic_callback (called for every received message)
    subscription_ = this->create_subscription<std_msgs::msg::String>(
      "chatter", 10,
      std::bind(&SubscriberNode::topic_callback, this, std::placeholders::_1));

    RCLCPP_INFO(this->get_logger(), "Subscriber node started, listening on /chatter");
  }

private:
  void topic_callback(const std_msgs::msg::String::SharedPtr msg) const
  {
    RCLCPP_INFO(this->get_logger(), "Received: '%s'", msg->data.c_str());
  }

  rclcpp::Subscription<std_msgs::msg::String>::SharedPtr subscription_;
};

int main(int argc, char * argv[])
{
  rclcpp::init(argc, argv);
  rclcpp::spin(std::make_shared<SubscriberNode>());
  rclcpp::shutdown();
  return 0;
}

File: src/pkg02_topic_py/pkg02_topic_py/subscriber_node.py
#!/usr/bin/env python3
"""
A minimal ROS2 subscriber in Python.
Subscribes to /chatter and logs every received message.
"""

import rclpy
from rclpy.node import Node
from std_msgs.msg import String


class SubscriberNode(Node):
    def __init__(self):
        super().__init__('subscriber_node_py')

        # Create subscriber. Same topic name & type as the publisher.
        self.subscription = self.create_subscription(
            String,
            'chatter',
            self.listener_callback,
            10)

        self.get_logger().info('Subscriber node (Python) listening on /chatter')

    def listener_callback(self, msg: String):
        self.get_logger().info(f'Received: "{msg.data}"')


def main(args=None):
    rclpy.init(args=args)
    node = SubscriberNode()
    try:
        rclpy.spin(node)
    except KeyboardInterrupt:
        pass
    finally:
        node.destroy_node()
        rclpy.shutdown()


if __name__ == '__main__':
    main()
